lib/orphan_importer.rb

```diff
@@ -1,52 +1,154 @@
 require_relative 'excel_upload'
-require_relative 'import_orphan_settings'
 
 class OrphanImporter
-  attr_accessor :import_errors, :settings
+
+#
+# This is the class which handles the importing of orphan rows from the spreadsheet. 
+# The rows are read in from the spreadsheet using the roo gem, and the information in each spreadsheet cell
+# is read using the roo gem #cell and #celltype methods. 
+# To make the rows ready for import into the database this OrphanImporter class does 4 things:
+#   1. It checks the celltype which roo returns to make sure that it aligns with the expected type for that cell.
+#      If it doesn't align, then it converts it to a String, and this may result in the later validation failing.
+#      Note that roo works out the type based on the type of data in the spreadsheet cell. For example, if the
+#      data in the cell is a number, roo will pass the type as Float. 
+#      See http://www.rubydoc.info/gems/roo/Roo/Excelx#celltype-instance_method
+#   2. It converts the value in some fields using the hashes below (eg to convert Arabic to English). 
+#      (This may result in an error - eg if the Arabic string is misspelt)
+#   3. It maps that spreadsheet cell to the appropriate orphan attribute, based on the spreadsheet column
+#   4. It validates the orphan attributes using the Rails orphan Model, and also checks that there aren't
+#      duplicate entries in the rows read in from the spreadsheet.
+# Steps 1 to 3 above are governed through the @@mapper class variable below.
+# If any errors are identified then details of these are added to the @import_errors array. 
+# The output from the extract_orphans method below is an array of 3 entries:
+#   - a boolean specifying if the spreadsheet was clean or not (ie no validation errors found)
+#   - an array of the (pending) orphan records
+#   - an array of the import errors identified
+#  
+  @@yn_to_bool = { 'y' => true, 'Y' => true, 'n' => false, 'N' => false }
+  @@male_arabic_yn_to_bool = { 'موجود' => true, 'متوفى' => false, 'y' => true, 'Y' => true, 'n' => false, 'N' => false }
+  @@female_arabic_yn_to_bool = { 'موجودة' => true, 'متوفاة' => false, 'y' => true, 'Y' => true, 'n' => false, 'N' => false }
+  @@gender = { 'ذكر' => 'Male', 'أنثى' => 'Female'}
+  @@province_to_code = { 'دمشق وريف دمشق' => 11, 'حلب' => 12, 'حمص' => 13, 'حماة' => 14, 'اللاذقية' => 15,
+    'دير الزور' => 16, 'درعا' => 17, 'إدلب' => 18, 'الرقة' => 19, 'الحسكة' => 20, 'طرطوس' => 21,
+    'السويداء' => 22, 'القنيطرة' => 23, 'خارج سوريا' => 29 } 
+  @@mapper = 
```

Comment:
Yes I've checked and it would be easy to change to drive the mapping off the column titles. (My only doubt was how it handled the merged cells.)
Nik, could you confirm with the clients that they'd want to go down this route? It would mean that they wouldn't be able to change the header titles in the first 3 rows without impacting the software. It might also then be better to have this mapping in a config file, rather than in a ruby file. Is there a natural place to put such a file within the hierarchy? What do you guys think?